docs/spec.md

```diff
@@ -1135,85 +1136,253 @@ in Evy. The call arguments must be listed individually.
-to any subtype. e.g. `[]` can be passed to a function that takes an
-argument of `[]num`, or `[]string`, etc.
+The `typeof` function returns the concrete type of a value held by a
+variable as a string. It returns a string that is the same as the type
+in an Evy program, such as `"num"`, `"bool"`, `"string"`, `"[]num"`, `"
+{}[]any"`, etc. It is particularly useful to determine the concrete
+type of an `any` variable together with type assertions.
 
-    typeof "abc"        // string
-    typeof true         // bool
-    arr := [ "abc" 1 ]
-    typeof arr          // []any
-    typeof arr[0]       // string
-    typeof arr[1]       // num
-    typeof {}           // {}
-    typeof []           // []
+Here is an example of how the `typeof` function works
 
-## Type assertion
+```evy
+print (typeof "abc")
+print (typeof true)
+print
+
+arr := ["abc" 1]
+print (typeof arr)
+print (typeof arr[0])
+print (typeof arr[1])
+```
+
+The output of this code is
+
+```evy:output
+string
+bool
+
+[]any
+string
+num
+```
+
+Empty composite literals, `[]` and `{}`, can be assigned to variables or
+parameters of any subtype, such as `[]string` or `{}num`. This is
+because empty composite literals are generic, meaning that they can be
+matched to any subtype.
+
+```evy
+func fn nums:[]num
+    print nums
+end
+
+fn []
+```
+
+The `typeof` functions will return `"[]"` or `"{}"` for an empty
+composite literal.
+
+An array literal, such as `[1 2 3]`, has a type of `[]num`. However, it
+is possible to assign an array literal of any type to a variable of
+type `[]any`. It is important to note that this only applies to array
+literals. A variable of type `[]num` cannot be assigned to a variable
+of type `[]any`.
+
+```evy
+x := [1 2 3]
+print "x" (typeof x)
+y:[]any
+// y = [1 2 3] // TODO PR https://github.com/foxygoat/evy/pull/179
+print "y" (typeof y)
+// y = x // compile time error
+// x = y // compile time error
+```
+
+will output
+
+```evy:output
+x []num
+y []any
+```
+
+## Type Assertion
 
 A type assertion `ident.(type)` asserts that the value of the variable
 `ident` is of the given `type`. If the assertion does not hold a
 [run-time panic](#run-time-panics-and-recoverable-errors) occurs.
 
-    x:any
-    x = [ 1 2 3 4 ]
-    num_array := x.([]num)
-    x = "abc"
-    str := x.(string)
+```evy
+x:any
+x = [1 2 3 4]
+num_array := x.([]num)
+print "typeof x:" (typeof x)
+print "typeof num_array" (typeof num_array)
+print
+
+x = "abc"
+str := x.(string)
+print "typeof x:" (typeof x)
+print "typeof str:" (typeof str)
+```
+
+Will generate the output
+
+```ev:output
+typeof x: any
+typeof num_array: []num
+
+typeof x: any
+typeof str: string
+```
 
 Only values of type `any` can be type asserted. That means an array of
-type any, `[]any`, _cannot_ be type assert to be an array of type `num`
-or other concrete type:
+type any, `[]any`, _cannot_ be type asserted to be an array of type
+`num` or any other concrete type. However, the elements of an array of
```

Comment:
`num` or `[]num` here? earlier in the sentence you say "an array of type any, `[]any`", then you say "an array of type `num`".

Reply:
yes, changed to `[]num`.